File: app/src/main/java/com/dougkeen/bart/backend/TransitFeedSnapshot.java
package com.dougkeen.bart.backend;

import com.google.transit.realtime.GtfsRealtime;

import java.util.Objects;

/**
 * One coherent read of the realtime feeds. The feeds are kept raw at this
 * boundary; route, station, and trip projections are built from this value.
 */
public final class TransitFeedSnapshot {
    private final GtfsRealtime.FeedMessage tripUpdates;
    private final GtfsRealtime.FeedMessage alerts;
    private final long receivedAtMillis;

    public TransitFeedSnapshot(GtfsRealtime.FeedMessage tripUpdates,
                               GtfsRealtime.FeedMessage alerts,
                               long receivedAtMillis) {
        this.tripUpdates = Objects.requireNonNull(tripUpdates, "tripUpdates");
        this.alerts = Objects.requireNonNull(alerts, "alerts");
        this.receivedAtMillis = receivedAtMillis;
    }

    public GtfsRealtime.FeedMessage getTripUpdates() {
        return tripUpdates;
    }

    public GtfsRealtime.FeedMessage getAlerts() {
        return alerts;
    }

    public long getReceivedAtMillis() {
        return receivedAtMillis;
    }

    public long getTripUpdatesTimestampMillis() {
        return feedTimestampMillis(tripUpdates, receivedAtMillis);
    }

    public long getAlertsTimestampMillis() {
        return feedTimestampMillis(alerts, receivedAtMillis);
    }

    /** Returns true when a new fetch contains no changed feed data. */
    public boolean hasSameFeedData(TransitFeedSnapshot other) {
        return other != null
                && tripUpdates.equals(other.tripUpdates)
                && alerts.equals(other.alerts);
    }

    private static long feedTimestampMillis(GtfsRealtime.FeedMessage feed,
                                             long fallback) {
        if (feed.hasHeader() && feed.getHeader().hasTimestamp()
                && feed.getHeader().getTimestamp() > 0) {
            return feed.getHeader().getTimestamp() * 1000L;
        }
        return fallback;
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/backend/TransitFeedSnapshot.kt
package com.dougkeen.bart.backend

import com.dougkeen.bart.networktasks.GtfsRealtimeFeedIndex
import com.google.transit.realtime.GtfsRealtime

/** One coherent read of the realtime feeds. */
class TransitFeedSnapshot(
    val tripUpdates: GtfsRealtime.FeedMessage,
    val alerts: GtfsRealtime.FeedMessage,
    val receivedAtMillis: Long
) {
    @Volatile
    private var tripUpdateIndex: GtfsRealtimeFeedIndex? = null

    @Volatile
    private var alertIndex: GtfsRealtimeFeedIndex? = null

    init {
        requireNotNull(tripUpdates) { "tripUpdates" }
        requireNotNull(alerts) { "alerts" }
    }

    fun getTripUpdatesTimestampMillis(): Long =
        feedTimestampMillis(tripUpdates, receivedAtMillis)

    fun getAlertsTimestampMillis(): Long =
        feedTimestampMillis(alerts, receivedAtMillis)

    /** Builds the trip-update index once and shares it across all projections. */
    fun getTripUpdateIndex(): GtfsRealtimeFeedIndex {
        tripUpdateIndex?.let { return it }
        synchronized(this) {
            tripUpdateIndex?.let { return it }
            return GtfsRealtimeFeedIndex.from(tripUpdates).also {
                tripUpdateIndex = it
            }
        }
    }

    /** Builds the alert index once and shares it across all projections. */
    fun getAlertIndex(): GtfsRealtimeFeedIndex {
        alertIndex?.let { return it }
        synchronized(this) {
            alertIndex?.let { return it }
            return GtfsRealtimeFeedIndex.from(alerts).also {
                alertIndex = it
            }
        }
    }

    /** Returns true when a new fetch contains no changed feed data. */
    fun hasSameFeedData(other: TransitFeedSnapshot?): Boolean =
        other != null && tripUpdates == other.tripUpdates && alerts == other.alerts

    private fun feedTimestampMillis(
        feed: GtfsRealtime.FeedMessage,
        fallback: Long
    ): Long = if (feed.hasHeader()
        && feed.header.hasTimestamp()
        && feed.header.timestamp > 0
    ) {
        feed.header.timestamp * 1000L
    } else {
        fallback
    }
}
